mongo/description/server.go

```diff
@@ -35,6 +35,7 @@ type Server struct {
 	AverageRTT            time.Duration
 	AverageRTTSet         bool
 	Compression           []string // compression methods returned by server
+	ConnectionID          *uint32
```

Comment:
Having `ConnectionID` as a public field in `description.Server` makes it accessible in SDAM monitoring events: `ServerDescriptionChangedEvent`, `ServerHeartbeatSucceededEvent`, and via the `description.Topology` in `TopologyDescriptionChangedEvent`.

Is that intentional? [DRIVERS-1897](https://jira.mongodb.org/browse/DRIVERS-1897) does not note those events.

`description.Server` represents the SDAM [Server Description data structure](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#serverdescription). 

`ConnectionID` may be natural to add to the server description. `ConnectionID` is a field in the `hello` / legacy hello response. The server description is parsed directly from a `hello` / legacy hello response.

OTOH each connection to a server has a (presumably unique) `ConnectionID`. `ConnectionID` describes the connection to the server, and is less of a property of the server.

If that seems like an unknown risk. An alternative could be to store `ConnectionID` as part of [HandshakeInformation](https://github.com/mongodb/mongo-go-driver/blob/33fac989d3a3f042cd94b5aa3400accc0fac04a3/x/mongo/driver/operation/hello.go#L274). That could then be used to set the `ConnectionID` directly on the `topology.connection`.

Reply:
Ah an interesting point. Making `ConnectionID` accessible from those monitoring events was not intentional (although `ServerHeartbeatSucceededEvent` is mentioned in DRIVERS-1897). I don't see much harm in that field now being available on any struct containing a `description.Server`, though. 

While `ConnectionID` is not listed in the spec's [server description data structure](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#serverdescription), `ConnectionID` is a description of the server's connection to the client, and is therefore, as I see it, a property of the server. There are a couple other fields on our `description.Server` that are not listed in that part of the spec: `Compression`, `CanonocialAddr`, `HeartbeatInterval`, `MaxBatchCount`, `MaxDocumentSize` etc.

Reply:
I think whether the `connectionId` value in the response from `hello` is distinct per network connection is critical to the correct behavior of the `ServerConnectionID` feature. If `connectionId` is distinct per connection, then there will potentially be many `connectionId`s per `Server` and reporting only the first seen `connectionId` will make correlating client-side events to server-side events impossible.

Does the `hello` response on each network connection to a server return a different `connectionId`?

Reply:
I don't think the `connectionId` value returned by the server is distinct per network connection. It represents the connection between the server and a particular client (which could be, as you mentioned, over numerous network connections). See the server's representation of an external `Client` [here](https://github.com/mongodb/mongo/blob/master/src/mongo/db/client.h#L184-L189); `connectionId` is a field associated with a particular client, not a connection. If the server were to restart, however, the `connectionId` would likely change. Not much to do there; events will just have a new `ServerConnectionID`.

Reply:
I think the `connectionId` is distinct per network connection based on [this example](https://github.com/kevinAlbs/go-bootstrap/blob/93edce46d9668f45a69e3f676d08dfa74defad3c/investigations/connection_id/main.go#L53).

My understanding is that the server does not have the concept of a driver client. It only knows about individual connections. I.e. the server cannot identify all connections from a given driver client. [client.h](https://github.com/mongodb/mongo/blob/master/src/mongo/db/client.h#L184-L189) may be using the term `client` to refer to a single network connection.

I think the open question is whether a `server.Description` is always tied to a network connection. I think it might be. In which case, I think it is OK to put a connectionID on `server.Description`.

Reply:
See most [recent comment](https://github.com/mongodb/mongo-go-driver/pull/724#pullrequestreview-754162663): I'm now storing `ServerConnectionID` on `HandshakeInformation` instead of `description.Server`.